lib/ansible/modules/crypto/openssl_certificate.py

```diff
@@ -0,0 +1,760 @@
+                            self.message.append(
+                                'Invalid keyUsage component (got %s, expected all of %s to be present)' % (str(extension).split(', '), keyUsage)
+                            )
+
+        def _validate_extendedKeyUsage():
+            if self.keyUsage:
+                for extension_idx in range(0, self.cert.get_extension_count()):
+                    extension = self.cert.get_extension(extension_idx)
+                    if extension.get_short_name() == 'extendedKeyUsage':
+                        extKeyUsage = [crypto_utils.extendedKeyUsageLong.get(keyUsage, keyUsage) for keyUsage in self.extendedKeyUsage]
+                        if (not self.extendedKeyUsage_strict and not all(x in str(extension).split(', ') for x in extKeyUsage)) or \
+                           (self.extendedKeyUsage_strict and not set(extKeyUsage) == set(str(extension).split(', '))):
+                            self.message.append(
+                                'Invalid extendedKeyUsage component (got %s, expected all of %s to be present)' % (str(extension).split(', '), extKeyUsage)
+                            )
+
+        def _validate_subjectAltName():
+            if self.subjectAltName:
+                for extension_idx in range(0, self.cert.get_extension_count()):
+                    extension = self.cert.get_extension(extension_idx)
+                    if extension.get_short_name() == 'subjectAltName':
+                        l_subjectAltName = [altname.replace('IP', 'IP Address') for altname in self.subjectAltName]
+                        if (not self.subjectAltName_strict and not all(x in str(extension).split(', ') for x in l_subjectAltName)) or \
+                           (self.subjectAltName_strict and not set(l_subjectAltName) == set(str(extension).split(', '))):
+                            self.message.append(
+                                'Invalid subjectAltName component (got %s, expected all of %s to be present)' % (str(extension).split(', '), l_subjectAltName)
+                            )
+
+        def _validate_notBefore():
+            if self.notBefore:
+                if self.cert.get_notBefore() != self.notBefore:
+                    self.message.append(
+                        'Invalid notBefore component (got %s, expected %s to be present)' % (self.cert.get_notBefore(), self.notBefore)
+                    )
+
+        def _validate_notAfter():
+            if self.notAfter:
+                if self.cert.get_notAfter() != self.notAfter:
+                    self.message.append(
+                        'Invalid notAfter component (got %s, expected %s to be present)' % (self.cert.get_notAfter(), self.notAfter)
+                    )
+
+        def _validate_valid_at():
+            if self.valid_at:
+                if not (self.valid_at >= self.cert.get_notBefore() and self.valid_at <= self.cert.get_notAfter()):
```

Comment:
Since I expect in both size ASN1.TIME format date, I expect the string comparison to be working all the time. Do you have specific case when it might not be the case?

I admit it is volunteer here to use pyOpenSSL as much as possible in order to avoid to deal with cryptography so we keep things simple where possible. I'll put the TODO, because futur us and developper will love us for that.